Mathlib/Tactic/Relation/Symm.lean

```diff
@@ -39,36 +40,102 @@ initialize registerBuiltinAttribute {
 }
 
-open Lean.Elab.Tactic in
+end Mathlib.Tactic
+
+open Mathlib.Tactic
+
+namespace Lean.Expr
+
+/--
+Internal implementation of `Lean.Expr.symm`, `Lean.MVarId.symm`, and the user-facing tactic.
+
+`tgt` should be of the form `a ~ b`, and is used to index the symm lemmas.
+
+`k lem args body` should calculate a result,
+given a candidate `symm` lemma `lem`, which will have type `∀ args, body`.
+
+In `Lean.Expr.symm` this result will be a new `Expr`,
+and in `Lean.MVarId.symm` and `Lean.MVarId.symmAt` this result will be a new goal.
+-/
+-- This function is rather opaque, but the design with a generic continuation `k`
+-- is necessary in order to factor out all the common requirements below.
+def symmAux (tgt : Expr) (k : Expr → Array Expr → Expr → MetaM α) : MetaM α := do
+  let .app (.app rel _) _ := tgt
+    | throwError "symmetry lemmas only apply to binary relations, not{indentExpr tgt}"
+  for lem in ← (symmExt.getState (← getEnv)).getMatch rel do
+    try
+      let lem ← mkConstWithFreshMVarLevels lem
+      let (args, _, body) ← withReducible <| forallMetaTelescopeReducing (← inferType lem)
+      return (← k lem args body)
+    catch _ => pure ()
+  throwError "no applicable symmetry lemma found for{indentExpr tgt}"
+
+/-- Given a term `e : a ~ b`, construct a term in `b ~ a` using `@[symm]` lemmas. -/
+def symm (e : Expr) : MetaM Expr := do
+  symmAux (← inferType e) fun lem args body => do
+    let .true ← isDefEq args.back e | failure
+    mkExpectedTypeHint (mkAppN lem args) (← instantiateMVars body)
+
+end Lean.Expr
+
+namespace Lean.MVarId
+
+/--
+Internal implementation of `Lean.MVarId.symm` and the user-facing tactic.
+
+`tgt` should be of the form `a ~ b`, and is used to index the symm lemmas.
+
+`k lem args body goal` should transform `goal` into a new goal,
+given a candidate `symm` lemma `lem`, which will have type `∀ args, body`.
+Depending on whether we are working on a hypothesis or a goal,
+`k` will internally use either `replace` or `assign`.
+-/
+def symmAux (tgt : Expr) (k : Expr → Array Expr → Expr → MVarId → MetaM MVarId) (g : MVarId) :
+    MetaM MVarId := do
+  tgt.symmAux fun lem args body => do
+    let g' ← k lem args body g
+    g'.setTag (← g.getTag)
+    return g'
+
+/-- Apply a symmetry lemma (i.e. marked with `@[symm]`) to a metavariable. -/
+def symm (g : MVarId) : MetaM MVarId := do
+  g.symmAux (← g.getType) fun lem args body g => do
+    let .true ← isDefEq (← g.getType) body | failure
+    g.assign (mkAppN lem args)
+    return args.back.mvarId!
+
+/-- Use a symmetry lemma (i.e. marked with `@[symm]`) to replace a hypothesis in a goal. -/
+def symmAt (h : FVarId) (g : MVarId) : MetaM MVarId := do
+  let h' ← (Expr.fvar h).symm
+  pure (← g.replace h (← inferType h') h').mvarId
+
+/-- For every hypothesis `h : a ~ b` where a `@[symm]` lemma is available,
+add a hypothesis `h.symm : b ~ a`. -/
+-- We could omit duplicate propositions, and avoid clobbering names, but don't for now.
```

Comment:
This seems like a pretty important limitation, you might even call it a bug. The name "saturate" implies that it works until fixpoint but if you iterate this the local context explodes.